File: src/platform.zig
const std = @import("std");
const builtin = @import("builtin");
const io_ctx = @import("io_ctx.zig");

pub const OperatingSystem = enum {
    linux,
    macos,

    pub fn current() OperatingSystem {
        return switch (builtin.os.tag) {
            .macos => .macos,
            else => .linux,
        };
    }

    pub fn name(self: OperatingSystem) []const u8 {
        return switch (self) {
            .linux => "linux",
            .macos => "darwin",
        };
    }

    /// Title-cased OS name, e.g. "Linux", "Darwin" — used by tools like trivy.
    pub fn titleName(self: OperatingSystem) []const u8 {
        return switch (self) {
            .linux => "Linux",
            .macos => "macOS",
        };
    }

    pub fn goName(self: OperatingSystem) []const u8 {
        return switch (self) {
            .linux => "linux-amd64",
            .macos => "darwin-amd64",
        };
    }

    /// OS name as used by ziglang.org downloads: "linux" or "macos".
    pub fn zigName(self: OperatingSystem) []const u8 {
        return switch (self) {
            .linux => "linux",
            .macos => "macos",
        };
    }
};

pub const Arch = enum {
    x86_64,
    aarch64,
    arm,
    i386,

    pub fn current() Arch {
        return switch (builtin.cpu.arch) {
            .aarch64 => .aarch64,
            .arm => .arm,
            .x86 => .i386,
            else => .x86_64,
        };
    }

    /// Name used in most Go-based tool URLs (amd64/arm64)
    pub fn goName(self: Arch) []const u8 {
        return switch (self) {
            .x86_64 => "amd64",
            .aarch64 => "arm64",
            .arm => "arm",
            .i386 => "386",
        };
    }

    /// Debian/Ubuntu package arch name
    pub fn debName(self: Arch) []const u8 {
        return switch (self) {
            .x86_64 => "amd64",
            .aarch64 => "arm64",
            .arm => "armhf",
            .i386 => "i386",
        };
    }

    /// Raw uname -m name
    pub fn unameName(self: Arch) []const u8 {
        return switch (self) {
            .x86_64 => "x86_64",
            .aarch64 => "aarch64",
            .arm => "armv7l",
            .i386 => "i686",
        };
    }

    /// Hybrid arch name used by tools like lazygit (x86_64 for Intel, arm64 for ARM)
    pub fn altName(self: Arch) []const u8 {
        return switch (self) {
            .x86_64 => "x86_64",
            .aarch64 => "arm64",
            .arm => "arm",
            .i386 => "386",
        };
    }
};

pub const Shell = enum {
    bash,
    zsh,
    fish,
    unknown,

    pub fn detect() Shell {
        const shell_env = @import("env.zig").getenv("SHELL") orelse return .unknown;
        const shell_name = std.fs.path.basename(shell_env);
        if (std.mem.eql(u8, shell_name, "bash")) return .bash;
        if (std.mem.eql(u8, shell_name, "zsh")) return .zsh;
        if (std.mem.eql(u8, shell_name, "fish")) return .fish;
        return .unknown;
    }

    pub fn name(self: Shell) []const u8 {
        return switch (self) {
            .bash => "bash",
            .zsh => "zsh",
            .fish => "fish",
            .unknown => "unknown",
        };
    }

    /// Name of centralized integration file
    pub fn integrationFileName(self: Shell) []const u8 {
        return switch (self) {
            .bash => "shell-integration.bash",
            .zsh => "shell-integration.zsh",
            .fish => "shell-integration.fish",
            .unknown => "shell-integration.sh",
        };
    }

    /// Shell-specific syntax to add a directory to PATH
    pub fn pathAddSyntax(self: Shell, dir: []const u8, allocator: std.mem.Allocator) ![]u8 {
        return switch (self) {
            .fish => std.fmt.allocPrint(allocator, "set -gx PATH {s} $PATH", .{dir}),
            else => std.fmt.allocPrint(allocator, "export PATH=\"{s}:$PATH\"", .{dir}),
        };
    }
};

pub const PackageManager = enum {
    pacman,
    apt,
    dnf,
    yum,
    zypper,
    apk,
    brew,
    flatpak,
    snap,
    unknown,

    /// Detect the primary native package manager in priority order
    pub fn detect() PackageManager {
        const native = [_]PackageManager{ .pacman, .apt, .dnf, .yum, .zypper, .apk };
        for (native) |pm| {
            if (pm.isAvailable()) return pm;
        }
        if (PackageManager.brew.isAvailable()) return .brew;
        if (PackageManager.flatpak.isAvailable()) return .flatpak;
        if (PackageManager.snap.isAvailable()) return .snap;
        return .unknown;
    }

    /// Check if a package manager binary exists in PATH
    pub fn isAvailable(self: PackageManager) bool {
        const cmd = self.command() orelse return false;
        // Use std.process to check
        const result = std.process.run(std.heap.page_allocator, io_ctx.get(), .{
            .argv = &.{ "sh", "-c", std.fmt.allocPrint(
                std.heap.page_allocator,
                "command -v {s}",
                .{cmd},
            ) catch return false },
        }) catch return false;
        std.heap.page_allocator.free(result.stdout);
        std.heap.page_allocator.free(result.stderr);
        return result.term == .exited and result.term.exited == 0;
    }

    pub fn command(self: PackageManager) ?[]const u8 {
        return switch (self) {
            .pacman => "pacman",
            .apt => "apt",
            .dnf => "dnf",
            .yum => "yum",
            .zypper => "zypper",
            .apk => "apk",
            .brew => "brew",
            .flatpak => "flatpak",
            .snap => "snap",
            .unknown => null,
        };
    }

    /// Install command prefix (without package name)
    pub fn installArgs(self: PackageManager) []const []const u8 {
        return switch (self) {
            .pacman => &.{ "sudo", "pacman", "-S", "--noconfirm" },
            .apt => &.{ "sudo", "apt-get", "install", "-y" },
            .dnf => &.{ "sudo", "dnf", "install", "-y" },
            .yum => &.{ "sudo", "yum", "install", "-y" },
            .zypper => &.{ "sudo", "zypper", "install", "-y" },
            .apk => &.{ "sudo", "apk", "add" },
            .brew => &.{ "brew", "install" },
            .flatpak => &.{ "flatpak", "install", "-y" },
            .snap => &.{ "snap", "install" },
            .unknown => &.{},
        };
    }

    /// Remove command prefix (without package name)
    pub fn removeArgs(self: PackageManager) []const []const u8 {
        return switch (self) {
            .pacman => &.{ "sudo", "pacman", "-R" },
            .apt => &.{ "sudo", "apt-get", "remove", "-y" },
            .dnf => &.{ "sudo", "dnf", "remove", "-y" },
            .yum => &.{ "sudo", "yum", "remove", "-y" },
            .zypper => &.{ "sudo", "zypper", "remove", "-y" },
            .apk => &.{ "sudo", "apk", "del" },
            .brew => &.{ "brew", "uninstall" },
            .flatpak => &.{ "flatpak", "uninstall", "-y" },
            .snap => &.{ "snap", "remove" },
            .unknown => &.{},
        };
    }
};

test "OperatingSystem.current does not panic" {
    _ = OperatingSystem.current();
}

test "Arch.current does not panic" {
    _ = Arch.current();
}

test "Shell.detect does not panic" {
    _ = Shell.detect();
}

test "PackageManager command names" {
    try std.testing.expectEqualStrings("pacman", PackageManager.pacman.command().?);
    try std.testing.expectEqualStrings("apt", PackageManager.apt.command().?);
}

File: src/platform.rs
use std::env;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OperatingSystem {
    Linux,
    Macos,
}

impl OperatingSystem {
    pub fn current() -> Self {
        if cfg!(target_os = "macos") {
            OperatingSystem::Macos
        } else {
            OperatingSystem::Linux
        }
    }

    pub fn name(self) -> &'static str {
        match self {
            OperatingSystem::Linux => "linux",
            OperatingSystem::Macos => "darwin",
        }
    }

    pub fn title_name(self) -> &'static str {
        match self {
            OperatingSystem::Linux => "Linux",
            OperatingSystem::Macos => "macOS",
        }
    }

    pub fn zig_name(self) -> &'static str {
        match self {
            OperatingSystem::Linux => "linux",
            OperatingSystem::Macos => "macos",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Arch {
    X86_64,
    Aarch64,
    Arm,
    I386,
}

impl Arch {
    pub fn current() -> Self {
        if cfg!(target_arch = "aarch64") {
            Arch::Aarch64
        } else if cfg!(target_arch = "arm") {
            Arch::Arm
        } else if cfg!(target_arch = "x86") {
            Arch::I386
        } else {
            Arch::X86_64
        }
    }

    pub fn go_name(self) -> &'static str {
        match self {
            Arch::X86_64 => "amd64",
            Arch::Aarch64 => "arm64",
            Arch::Arm => "arm",
            Arch::I386 => "386",
        }
    }

    pub fn uname_name(self) -> &'static str {
        match self {
            Arch::X86_64 => "x86_64",
            Arch::Aarch64 => "aarch64",
            Arch::Arm => "armv7l",
            Arch::I386 => "i686",
        }
    }

    pub fn alt_name(self) -> &'static str {
        match self {
            Arch::X86_64 => "x86_64",
            Arch::Aarch64 => "arm64",
            Arch::Arm => "arm",
            Arch::I386 => "386",
        }
    }

    pub fn x64_name(self) -> &'static str {
        match self {
            Arch::X86_64 => "x64",
            Arch::Aarch64 => "arm64",
            Arch::Arm => "arm",
            Arch::I386 => "x86",
        }
    }

    pub fn rust_target(self, os: OperatingSystem) -> &'static str {
        match os {
            OperatingSystem::Linux => match self {
                Arch::X86_64 => "x86_64-unknown-linux-gnu",
                Arch::Aarch64 => "aarch64-unknown-linux-gnu",
                Arch::Arm => "armv7-unknown-linux-gnueabihf",
                Arch::I386 => "i686-unknown-linux-gnu",
            },
            OperatingSystem::Macos => match self {
                Arch::X86_64 => "x86_64-apple-darwin",
                Arch::Aarch64 => "aarch64-apple-darwin",
                Arch::Arm => "arm-apple-darwin",
                Arch::I386 => "i686-apple-darwin",
            },
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Shell {
    Bash,
    Zsh,
    Fish,
    Unknown,
}

impl Shell {
    pub fn detect() -> Self {
        let shell_env = match env::var("SHELL") {
            Ok(v) => v,
            Err(_) => return Shell::Unknown,
        };
        let name = std::path::Path::new(&shell_env)
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("");
        match name {
            "bash" => Shell::Bash,
            "zsh" => Shell::Zsh,
            "fish" => Shell::Fish,
            _ => Shell::Unknown,
        }
    }

    pub fn name(self) -> &'static str {
        match self {
            Shell::Bash => "bash",
            Shell::Zsh => "zsh",
            Shell::Fish => "fish",
            Shell::Unknown => "unknown",
        }
    }

    pub fn integration_file_name(self) -> &'static str {
        match self {
            Shell::Bash => "shell-integration.bash",
            Shell::Zsh => "shell-integration.zsh",
            Shell::Fish => "shell-integration.fish",
            Shell::Unknown => "shell-integration.sh",
        }
    }

    pub fn path_add_syntax(self, dir: &str) -> String {
        match self {
            Shell::Fish => format!("set -gx PATH {dir} $PATH"),
            _ => format!("export PATH=\"{dir}:$PATH\""),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PackageManager {
    Pacman,
    Apt,
    Dnf,
    Yum,
    Zypper,
    Apk,
    Brew,
    Flatpak,
    Snap,
    Unknown,
}

impl PackageManager {
    pub fn detect() -> Self {
        let native = [
            PackageManager::Pacman,
            PackageManager::Apt,
            PackageManager::Dnf,
            PackageManager::Yum,
            PackageManager::Zypper,
            PackageManager::Apk,
        ];
        for pm in native {
            if pm.is_available() {
                return pm;
            }
        }
        if PackageManager::Brew.is_available() {
            return PackageManager::Brew;
        }
        if PackageManager::Flatpak.is_available() {
            return PackageManager::Flatpak;
        }
        if PackageManager::Snap.is_available() {
            return PackageManager::Snap;
        }
        PackageManager::Unknown
    }

    pub fn is_available(self) -> bool {
        let cmd = match self.command() {
            Some(c) => c,
            None => return false,
        };
        crate::util::find_in_path(cmd).is_some()
    }

    pub fn command(self) -> Option<&'static str> {
        match self {
            PackageManager::Pacman => Some("pacman"),
            PackageManager::Apt => Some("apt"),
            PackageManager::Dnf => Some("dnf"),
            PackageManager::Yum => Some("yum"),
            PackageManager::Zypper => Some("zypper"),
            PackageManager::Apk => Some("apk"),
            PackageManager::Brew => Some("brew"),
            PackageManager::Flatpak => Some("flatpak"),
            PackageManager::Snap => Some("snap"),
            PackageManager::Unknown => None,
        }
    }

    pub fn install_args(self) -> &'static [&'static str] {
        match self {
            PackageManager::Pacman => &["sudo", "pacman", "-S", "--noconfirm"],
            PackageManager::Apt => &["sudo", "apt-get", "install", "-y"],
            PackageManager::Dnf => &["sudo", "dnf", "install", "-y"],
            PackageManager::Yum => &["sudo", "yum", "install", "-y"],
            PackageManager::Zypper => &["sudo", "zypper", "install", "-y"],
            PackageManager::Apk => &["sudo", "apk", "add"],
            PackageManager::Brew => &["brew", "install"],
            PackageManager::Flatpak => &["flatpak", "install", "-y"],
            PackageManager::Snap => &["snap", "install"],
            PackageManager::Unknown => &[],
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn os_current_does_not_panic() {
        let _ = OperatingSystem::current();
    }

    #[test]
    fn arch_current_does_not_panic() {
        let _ = Arch::current();
    }

    #[test]
    fn shell_detect_does_not_panic() {
        let _ = Shell::detect();
    }

    #[test]
    fn package_manager_command_names() {
        assert_eq!(PackageManager::Pacman.command(), Some("pacman"));
        assert_eq!(PackageManager::Apt.command(), Some("apt"));
        assert_eq!(PackageManager::Unknown.command(), None);
    }

    #[test]
    fn rust_target_linux_x86_64() {
        assert_eq!(
            Arch::X86_64.rust_target(OperatingSystem::Linux),
            "x86_64-unknown-linux-gnu"
        );
    }

    #[test]
    fn rust_target_macos_aarch64() {
        assert_eq!(
            Arch::Aarch64.rust_target(OperatingSystem::Macos),
            "aarch64-apple-darwin"
        );
    }
}
